src/tls.c

```diff
@@ -1579,134 +1579,190 @@ static int TLSX_SetALPN(TLSX** extensions, const void* data, word16 size,
     return WOLFSSL_SUCCESS;
 }
 
-/** Parses a buffer of ALPN extensions and set the first one matching
- * client and server requirements */
-static int TLSX_ALPN_ParseAndSet(WOLFSSL *ssl, const byte *input, word16 length,
-                                 byte isRequest)
+static int ALPN_find_match(WOLFSSL *ssl, TLSX **pextension,
+                           const byte **psel, byte *psel_len,
+                           const byte *alpn_val, word16 alpn_val_len)
 {
-    word16  size = 0, offset = 0, idx = 0;
-    int     r = BUFFER_ERROR;
-    byte    match = 0;
     TLSX    *extension;
-    ALPN    *alpn = NULL, *list;
-
-    if (OPAQUE16_LEN > length)
-        return BUFFER_ERROR;
-
-    ato16(input, &size);
-    offset += OPAQUE16_LEN;
-
-    if (size == 0)
-        return BUFFER_ERROR;
+    ALPN    *alpn, *list;
+    int     r = 0;
+    const byte *sel = NULL, *s;
+    byte sel_len = 0, wlen;
 
     extension = TLSX_Find(ssl->extensions, TLSX_APPLICATION_LAYER_PROTOCOL);
     if (extension == NULL)
         extension = TLSX_Find(ssl->ctx->extensions,
                               TLSX_APPLICATION_LAYER_PROTOCOL);
 
-#if defined(OPENSSL_ALL) || defined(WOLFSSL_NGINX) || defined(WOLFSSL_HAPROXY)
-    if (ssl->alpnSelect != NULL && ssl->options.side == WOLFSSL_SERVER_END) {
-        const byte* out;
-        unsigned char outLen;
-
-        if (ssl->alpnSelect(ssl, &out, &outLen, input + offset, size,
-                            ssl->alpnSelectArg) == 0) {
-            WOLFSSL_MSG("ALPN protocol match");
-            /* clears out all current ALPN extensions set */
-            TLSX_Remove(&ssl->extensions, TLSX_APPLICATION_LAYER_PROTOCOL, ssl->heap);
-            extension = NULL;
-            if (TLSX_UseALPN(&ssl->extensions, (char*)out, outLen, 0, ssl->heap)
-                                                           == WOLFSSL_SUCCESS) {
-                extension = TLSX_Find(ssl->extensions,
-                                      TLSX_APPLICATION_LAYER_PROTOCOL);
-            }
-        }
-    }
-#endif
-
+    /* No ALPN configured here */
     if (extension == NULL || extension->data == NULL) {
-        return isRequest ? 0
-                         : TLSX_HandleUnsupportedExtension(ssl);
+        extension = NULL;
+        goto cleanup;
```

Comment:
No new gotos in the code. Part of our coding standard. Please refactor.

Reply:
Ah, was not aware of that law. Removed the gotos again.